editor/src/core/shared/element-path-tree.ts

```diff
@@ -133,6 +134,73 @@ export function reorderTree(trees: ElementPathTrees, metadata: ElementInstanceMe
   }
 }
 
+function maybeReorderDynamicChildren(
+  children: ElementPathTree[],
+  metadata: ElementInstanceMetadataMap,
+): ElementPathTree[] {
+  function keepOnlyUnique(path: ElementPath, index: number, array: ElementPath[]) {
+    return array.findIndex((e) => EP.pathsEqual(path, e)) === index
+  }
+
+  // Get all the dynamic children and sort them
+  // alphabetically, since their paths will have incremental indexes (~~~N).
+  const dynamicChildren = children
+    .filter((child) => EP.isDynamicPath(child.path))
+    .sort((a, b) => a.pathString.localeCompare(b.pathString))
+
+  // If there are no dynamic children, do nothing.
+  if (dynamicChildren.length === 0) {
+    return children
+  }
+
+  const metadataKeys = Object.keys(metadata)
+
+  // Build a stack to reorder dynamic children
+  const dynamicChildrenStack = dynamicChildren
+    .sort((a, b) => {
+      return metadataKeys.indexOf(a.pathString) - metadataKeys.indexOf(b.pathString)
+    })
```

Comment:
What I don't like about this part is that this is essentially `O(metadataKeys.length * dynamicChildren.length)`, and ideally we'd like to avoid depending on project size for performance. This can be easily fixed by calculating a reverse lookup for indices up front and using it here:
```typescript
const metadataKeysToIndices = Object.keys(metadata).reduce(
    (acc: { [key: string]: number }, val: string, idx: number) => {
      acc[val] = idx
      return acc
    },
    {},
  )
  ```
  
```typescript
  // Build a stack to reorder dynamic children
  const dynamicChildrenStack = dynamicChildren
    .sort((a, b) => {
      return metadataKeys[a.pathString] - metadataKeys[b.pathString]
    })
    .map((c) => EP.dynamicPathToStaticPath(c.path))
    .filter(keepOnlyUnique)
 ```